### DISCUSSION TITLE
Is there any drawback of writing a custom high-order hook for `useAtom`?

### DISCUSSION THREAD

**Discussion by rodrigocfd** [pmndrs/jotai#1228] [Q&A] [answered]
Imagine this hypothetical [custom hook](https://reactjs.org/docs/hooks-custom.html), which wraps `useAtom`:

```ts
import {atom, useAtom} from 'jotai';

const nameAtom = atom('Joe'); // note: atom itself won't be exported

export function useName() { // high-order custom hook
    const [name, setName] = useAtom();
    return {
        value: name, // simplest getter
        addSurname(surname: string) { // modifier
            setName(name + ' ' + surname);
        },
        upper() { // modifier
            setName(name.toUpperCase());
        },
    };
}
```
With the custom hook above, updates would be a lot easier than [write only atoms](https://github.com/pmndrs/jotai#write-only-atoms), at least to my tired eyes:

```ts
import {useName} from './useName';

function Person() {
    const name = useName();

    return <div>
        <h1>{name.value}</h1>
        <button onClick={() => name.addSurname('Foo')}>Add surname</button>
        <button onClick={() => name.upper()}>Upper</button>
    </div>;
}

export default Person;
```

So far, for all the tests I've written, it works flawlessly... is there any drawback of using this approach, like a memory leak, faulty update, dealocks, etc?

Is this approach safe to be used?

**Comment by dai-shi** [comment, marked as answer]
No, it's totally fine. What you showed is safe to use.

I would say it's mostly preference to use write-only atoms or custom hooks. A small drawback with custom hooks is you might need memoizing callbacks. Or, `useEvent` that will be available in the future. With atoms, it's already handled. (for example: https://twitter.com/dai_shi/status/1522205634724524032 )
